File: TarsEngine/Services/FSharpStructureExtractor.cs
using System.Text.RegularExpressions;
using Microsoft.Extensions.Logging;
using TarsEngine.Models;
using TarsEngine.Services.Interfaces;

namespace TarsEngine.Services;

/// <summary>
/// Extracts code structures from F# code
/// </summary>
public class FSharpStructureExtractor(ILogger<FSharpStructureExtractor> logger) : ICodeStructureExtractor
{
    private readonly ILogger<FSharpStructureExtractor> _logger = logger;

    /// <inheritdoc/>
    public string Language => "fsharp";

    /// <inheritdoc/>
    public List<CodeStructure> ExtractStructures(string content)
    {
        var structures = new List<CodeStructure>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return structures;
            }

            // Extract modules
            var moduleRegex = new Regex(@"module\s+([a-zA-Z0-9_.]+)(?:\s*=)?", RegexOptions.Compiled);
            var moduleMatches = moduleRegex.Matches(content);
            foreach (Match match in moduleMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var moduleName = match.Groups[1].Value;
                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Module,
                        Name = moduleName,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = moduleName
                        }
                    });
                }
            }

            // Extract types (records, discriminated unions, classes)
            var typeRegex = new Regex(@"type\s+([a-zA-Z0-9_]+)(?:<[^>]+>)?(?:\s*=|\s*\(|\s*:)", RegexOptions.Compiled);
            var typeMatches = typeRegex.Matches(content);
            foreach (Match match in typeMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var typeName = match.Groups[1].Value;
                    var moduleName = GetModuleForPosition(structures, match.Index, content);

                    // Determine type kind (record, union, class)
                    var typeKind = DetermineTypeKind(content, match.Index);

                    structures.Add(new CodeStructure
                    {
                        Type = typeKind,
                        Name = typeName,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = moduleName,
                            ClassName = typeName
                        },
                        Properties = new Dictionary<string, string>
                        {
                            { "TypeKind", typeKind.ToString() }
                        }
                    });
                }
            }

            // Extract functions
            var functionRegex = new Regex(@"let\s+(rec\s+)?([a-zA-Z0-9_]+)(?:\s+[a-zA-Z0-9_]+)*\s*=", RegexOptions.Compiled);
            var functionMatches = functionRegex.Matches(content);
            foreach (Match match in functionMatches)
            {
                if (match.Groups.Count > 2)
                {
                    var functionName = match.Groups[2].Value;
                    var moduleName = GetModuleForPosition(structures, match.Index, content);
                    var isRecursive = !string.IsNullOrEmpty(match.Groups[1].Value);

                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Function,
                        Name = functionName,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = moduleName
                        },
                        Properties = new Dictionary<string, string>
                        {
                            { "IsRecursive", isRecursive.ToString() }
                        }
                    });
                }
            }

            // Extract members
            var memberRegex = new Regex(@"member\s+(?:this|self|_)\.([a-zA-Z0-9_]+)", RegexOptions.Compiled);
            var memberMatches = memberRegex.Matches(content);
            foreach (Match match in memberMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var memberName = match.Groups[1].Value;
                    var moduleName = GetModuleForPosition(structures, match.Index, content);
                    var className = GetClassForPosition(structures, match.Index, content);

                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Method,
                        Name = memberName,
                        ParentName = className,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = moduleName,
                            ClassName = className
                        }
                    });
                }
            }

            // Extract active patterns
            var activePatternRegex = new Regex(@"\(\|([^|]+)\|\)", RegexOptions.Compiled);
            var activePatternMatches = activePatternRegex.Matches(content);
            foreach (Match match in activePatternMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var patternName = match.Groups[1].Value;
                    var moduleName = GetModuleForPosition(structures, match.Index, content);

                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Function,
                        Name = patternName,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = moduleName
                        },
                        Properties = new Dictionary<string, string>
                        {
                            { "IsActivePattern", "true" }
                        }
                    });
                }
            }

            // Calculate structure sizes and update end lines
            CalculateStructureSizes(structures, content);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error extracting structures from F# code");
        }

        return structures;
    }

    /// <inheritdoc/>
    public string GetNamespaceForPosition(List<CodeStructure> structures, int position, string content)
    {
        // In F#, modules are equivalent to namespaces
        var lineNumber = GetLineNumber(content, position);
        var moduleStructures = structures
            .Where(s => s.Type == StructureType.Module)
            .ToList();

        foreach (var module in moduleStructures)
        {
            if (module.Location.StartLine <= lineNumber &&
                (module.Location.EndLine == 0 || module.Location.EndLine >= lineNumber))
            {
                return module.Name;
            }
        }

        return string.Empty;
    }

    /// <inheritdoc/>
    public string GetClassForPosition(List<CodeStructure> structures, int position, string content)
    {
        var lineNumber = GetLineNumber(content, position);
        var classStructures = structures
            .Where(s => s.Type == StructureType.Class || s.Type == StructureType.Record || s.Type == StructureType.Union)
            .ToList();

        foreach (var cls in classStructures)
        {
            if (cls.Location.StartLine <= lineNumber &&
                (cls.Location.EndLine == 0 || cls.Location.EndLine >= lineNumber))
            {
                return cls.Name;
            }
        }

        return string.Empty;
    }

    /// <inheritdoc/>
    public void CalculateStructureSizes(List<CodeStructure> structures, string content)
    {
        // F# doesn't use braces for scope, so we need to use indentation to determine structure boundaries
        var lines = content.Split('\n');

        // Sort structures by start line
        var sortedStructures = structures.OrderBy(s => s.Location.StartLine).ToList();

        for (int i = 0; i < sortedStructures.Count; i++)
        {
            var structure = sortedStructures[i];
            var startLine = structure.Location.StartLine;

            if (startLine <= 0 || startLine > lines.Length)
            {
                continue;
            }

            // Get the indentation of the structure definition
            var structureIndent = GetIndentation(lines[startLine - 1]);

            // Find the end line by looking for the next line with the same or less indentation
            var endLine = startLine;
            for (int j = startLine; j < lines.Length; j++)
            {
                var line = lines[j];

                // Skip empty lines
                if (string.IsNullOrWhiteSpace(line))
                {
                    continue;
                }

                var indent = GetIndentation(line);

                // If we find a line with the same or less indentation, we've reached the end of the structure
                if (indent <= structureIndent && j > startLine)
                {
                    endLine = j - 1;
                    break;
                }

                // If we reach the end of the file, use that as the end line
                if (j == lines.Length - 1)
                {
                    endLine = j;
                }
            }

            // Update the structure
            structure.Location.EndLine = endLine;
            structure.Size = endLine - startLine + 1;

            // Check if there's a next structure that starts before this one ends
            if (i < sortedStructures.Count - 1)
            {
                var nextStructure = sortedStructures[i + 1];
                if (nextStructure.Location.StartLine <= structure.Location.EndLine)
                {
                    // Adjust the end line to be just before the next structure
                    structure.Location.EndLine = nextStructure.Location.StartLine - 1;
                    structure.Size = structure.Location.EndLine - structure.Location.StartLine + 1;
                }
            }

            // Ensure size is at least 1
            if (structure.Size <= 0)
            {
                structure.Size = 1;
            }
        }
    }

    /// <summary>
    /// Gets the line number for a position in the content
    /// </summary>
    public int GetLineNumber(string content, int position)
    {
        if (string.IsNullOrEmpty(content) || position < 0 || position >= content.Length)
        {
            return 0;
        }

        // Count newlines before the position
        return content[..position].Count(c => c == '\n') + 1;
    }

    /// <summary>
    /// Gets the module for a position in the content
    /// </summary>
    private string GetModuleForPosition(List<CodeStructure> structures, int position, string content)
    {
        var lineNumber = GetLineNumber(content, position);
        var moduleStructures = structures
            .Where(s => s.Type == StructureType.Module)
            .ToList();

        foreach (var module in moduleStructures)
        {
            if (module.Location.StartLine <= lineNumber &&
                (module.Location.EndLine == 0 || module.Location.EndLine >= lineNumber))
            {
                return module.Name;
            }
        }

        return string.Empty;
    }

    /// <summary>
    /// Determines the type kind (record, union, class) based on the content
    /// </summary>
    private StructureType DetermineTypeKind(string content, int position)
    {
        // Get the content after the type definition
        var afterType = content.Substring(position);

        // Check for record type
        if (afterType.Contains("{") && afterType.Contains(":"))
        {
            return StructureType.Record;
        }

        // Check for discriminated union
        if (afterType.Contains("|"))
        {
            return StructureType.Union;
        }

        // Check for class
        if (afterType.Contains("class") || afterType.Contains("new()"))
        {
            return StructureType.Class;
        }

        // Default to class
        return StructureType.Class;
    }

    /// <summary>
    /// Gets the indentation level of a line
    /// </summary>
    private static int GetIndentation(string line)
    {
        int indent = 0;
        foreach (char c in line)
        {
            if (c == ' ')
            {
                indent++;
            }
            else if (c == '\t')
            {
                indent += 4; // Assuming tabs are 4 spaces
            }
            else
            {
                break;
            }
        }
        return indent;
    }
}

File: TarsEngine.FSharp.Core/Analysis/FSharpStructureExtractor.fs
namespace TarsEngine.FSharp.Core.Analysis

open System
open System.Collections.Generic
open System.IO
open System.Text.RegularExpressions
open Microsoft.Extensions.Logging

/// <summary>
/// Implementation of ICodeStructureExtractor for F# language.
/// </summary>
type FSharpStructureExtractor(logger: ILogger<FSharpStructureExtractor>) =
    
    /// <summary>
    /// Gets the language supported by this extractor.
    /// </summary>
    member _.Language = "fsharp"
    
    /// <summary>
    /// Extracts code structures from the provided content.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of extracted code structures.</returns>
    member _.ExtractStructures(content: string) =
        try
            logger.LogInformation("Extracting structures from F# code")
            
            // Define regex patterns for different structure types
            let namespacePattern = @"namespace\s+([a-zA-Z0-9_\.]+)"
            let modulePattern = @"module\s+(?:rec\s+)?([a-zA-Z0-9_\.]+)"
            let typePattern = @"type\s+(?:private\s+)?([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*(?:=\s*[^{]+)?\s*(?:\{|=|\()"
            let recordPattern = @"type\s+(?:private\s+)?([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*=\s*\{[^}]*\}"
            let unionPattern = @"type\s+(?:private\s+)?([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*=(?:\s*\|)?\s*([^=]+)(?:\s*\|[^=]+)*"
            let functionPattern = @"let\s+(?:rec\s+)?(?:private\s+)?([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*(?:\([^)]*\))?\s*(?::\s*[^=]+)?\s*="
            let valuePattern = @"let\s+(?:mutable\s+)?(?:private\s+)?([a-zA-Z0-9_]+)\s*(?::\s*[^=]+)?\s*="
            let memberPattern = @"member\s+(?:private\s+)?(?:this|self|_)\.([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*(?:\([^)]*\))?\s*(?::\s*[^=]+)?\s*="
            let propertyPattern = @"member\s+(?:private\s+)?(?:this|self|_)\.([a-zA-Z0-9_]+)\s*(?::\s*[^=]+)?\s*with\s+(?:get|set)"
            let interfacePattern = @"interface\s+([a-zA-Z0-9_\.]+)(?:<[^>]+>)?"
            let openPattern = @"open\s+([a-zA-Z0-9_\.]+)"
            let attributePattern = @"\[<([a-zA-Z0-9_]+)(?:\(.*\))?\s*>\]"
            
            // Extract structures using regex
            let extractStructures pattern structureType =
                Regex.Matches(content, pattern, RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m -> 
                    let startLine = content.Substring(0, m.Index).Split('\n').Length
                    let endLine = startLine + m.Value.Split('\n').Length - 1
                    let modifiers = 
                        if m.Value.Contains("private") then ["private"]
                        elif m.Value.Contains("internal") then ["internal"]
                        else ["public"]
                    
                    let name = 
                        match structureType with
                        | "namespace" | "module" | "type" | "record" | "union" | "function" | "value" | "member" | "property" | "interface" | "open" | "attribute" ->
                            m.Groups.[1].Value
                        | _ -> ""
                    
                    let returnType = 
                        let returnTypeMatch = Regex.Match(m.Value, @":\s*([^=]+)")
                        if returnTypeMatch.Success then
                            Some (returnTypeMatch.Groups.[1].Value.Trim())
                        else
                            None
                    
                    let parameters = 
                        match structureType with
                        | "function" | "member" ->
                            let paramMatch = Regex.Match(m.Value, @"\(([^)]*)\)")
                            if paramMatch.Success then
                                let paramString = paramMatch.Groups.[1].Value
                                if String.IsNullOrWhiteSpace(paramString) then
                                    []
                                else
                                    paramString.Split(',')
                                    |> Array.map (fun p -> 
                                        let parts = p.Trim().Split(':')
                                        if parts.Length >= 2 then
                                            (parts.[0].Trim(), parts.[1].Trim())
                                        else
                                            ("", p.Trim())
                                    )
                                    |> Array.toList
                            else
                                []
                        | _ -> []
                    
                    {
                        Name = name
                        StructureType = structureType
                        Parent = None
                        Children = []
                        StartLine = startLine
                        EndLine = endLine
                        Modifiers = modifiers
                        ReturnType = returnType
                        Parameters = parameters
                        Properties = Map.empty
                    }
                )
                |> Seq.toList
            
            // Extract all structures
            let namespaces = extractStructures namespacePattern "namespace"
            let modules = extractStructures modulePattern "module"
            let types = extractStructures typePattern "type"
            let records = extractStructures recordPattern "record"
            let unions = extractStructures unionPattern "union"
            let functions = extractStructures functionPattern "function"
            let values = extractStructures valuePattern "value"
            let members = extractStructures memberPattern "member"
            let properties = extractStructures propertyPattern "property"
            let interfaces = extractStructures interfacePattern "interface"
            let opens = extractStructures openPattern "open"
            let attributes = extractStructures attributePattern "attribute"
            
            // Combine all structures
            let allStructures = 
                List.concat [
                    namespaces
                    modules
                    types
                    records
                    unions
                    functions
                    values
                    members
                    properties
                    interfaces
                    opens
                    attributes
                ]
            
            // Build parent-child relationships
            let structureMap = Dictionary<string, CodeStructure>()
            
            // Add structures to map
            for structure in allStructures do
                let key = $"{structure.StructureType}:{structure.Name}"
                if not (structureMap.ContainsKey(key)) then
                    structureMap.Add(key, structure)
            
            // Build parent-child relationships
            let structures = 
                allStructures
                |> List.map (fun structure ->
                    // Find parent
                    let parent = 
                        allStructures
                        |> List.filter (fun s -> 
                            s.StartLine < structure.StartLine && 
                            s.EndLine > structure.EndLine && 
                            s.StructureType <> "open" && 
                            s.StructureType <> "attribute"
                        )
                        |> List.sortByDescending (fun s -> s.StartLine)
                        |> List.tryHead
                    
                    // Find children
                    let children = 
                        allStructures
                        |> List.filter (fun s -> 
                            s.StartLine > structure.StartLine && 
                            s.EndLine < structure.EndLine && 
                            s.StructureType <> "open" && 
                            s.StructureType <> "attribute"
                        )
                        |> List.filter (fun s ->
                            // Ensure the child is a direct child
                            not (allStructures |> List.exists (fun p -> 
                                p.StartLine < s.StartLine && 
                                p.EndLine > s.EndLine && 
                                p.StartLine > structure.StartLine && 
                                p.EndLine < structure.EndLine
                            ))
                        )
                    
                    // Update structure with parent and children
                    { structure with 
                        Parent = parent
                        Children = children
                    }
                )
            
            structures
        with
        | ex ->
            logger.LogError(ex, "Error extracting structures from F# code")
            []
    
    /// <summary>
    /// Extracts code structures from a file.
    /// </summary>
    /// <param name="filePath">The path to the file.</param>
    /// <returns>A list of extracted code structures.</returns>
    member this.ExtractStructuresFromFile(filePath: string) =
        try
            logger.LogInformation("Extracting structures from F# file: {FilePath}", filePath)
            
            // Read the file content
            let content = File.ReadAllText(filePath)
            
            // Extract structures from the content
            this.ExtractStructures(content)
        with
        | ex ->
            logger.LogError(ex, "Error extracting structures from F# file: {FilePath}", filePath)
            []
    
    /// <summary>
    /// Gets a structure by name.
    /// </summary>
    /// <param name="structures">The list of structures to search.</param>
    /// <param name="name">The name of the structure to find.</param>
    /// <returns>The found structure, if any.</returns>
    member _.GetStructureByName(structures: CodeStructure list, name: string) =
        structures |> List.tryFind (fun s -> s.Name = name)
    
    /// <summary>
    /// Gets structures by type.
    /// </summary>
    /// <param name="structures">The list of structures to search.</param>
    /// <param name="structureType">The type of structures to find.</param>
    /// <returns>The list of found structures.</returns>
    member _.GetStructuresByType(structures: CodeStructure list, structureType: string) =
        structures |> List.filter (fun s -> s.StructureType = structureType)
    
    interface ICodeStructureExtractor with
        member this.Language = this.Language
        member this.ExtractStructures(content) = this.ExtractStructures(content)
        member this.ExtractStructuresFromFile(filePath) = this.ExtractStructuresFromFile(filePath)
        member this.GetStructureByName(structures, name) = this.GetStructureByName(structures, name)
        member this.GetStructuresByType(structures, structureType) = this.GetStructuresByType(structures, structureType)
